### PR TITLE
[SMTChecker] Make mergeVariables deterministic

### LINKED ISSUE #6411: [SMTChecker] Remove potential nondeterminism in mergeVariables
The set should be sorted by AST id to be safe against nondeterminism.

### PR DESCRIPTION
**Pull request by leonardoalt**
Fixes #6411

### CODE REVIEW

**Review by ekpyron — DISMISSED** [review]
Well, theoretically it'd be better to use that comparison all the way up to ``VariableUsage``... maybe we should even add a custom ``ASTNodePtrSet`` that compares by ID and use that... but fine for now I guess!

File: libsolidity/formal/SMTChecker.cpp [hunk 1]
```diff
@@ -1711,7 +1711,11 @@ TypePointer SMTChecker::typeWithoutPointer(TypePointer const& _type)
 
 void SMTChecker::mergeVariables(set<VariableDeclaration const*> const& _variables, smt::Expression const& _condition, VariableIndices const& _indicesEndTrue, VariableIndices const& _indicesEndFalse)
 {
-	for (auto const* decl: _variables)
+	auto cmp = [&] (VariableDeclaration const* var1, VariableDeclaration const* var2) {
```
Comment (ekpyron) [inline libsolidity/formal/SMTChecker.cpp]: And actually: You don't need to capture anything here...

**Comment by leonardoalt** [comment]
I would argue the opposite, that the only place that we care about the order of the touched variables is inside `mergeVariables` since it adds assertions.

**Comment by ekpyron** [comment]
But now you have to sort in ``mergeVariables`` - whereas if you use ID comparison everywhere (instead the default pointer comparison), it will already be sorted... and outside ``mergeVariables`` it doesn't matter which comparison is used for sorting, so why not sort by ID all the time? One could argue that sorting by ID means one more indirection while inserting... But yeah, it really doesn't matter :-D.

**Review by ekpyron — APPROVED** [review]

**Pull request merged**